### PR TITLE
cover/cmds/core/cat 

### PR DESCRIPTION
**Pull request by ChriMarMe**
Restructure main
Add test for empty arguments
Add test for file not exist
Add test for io copy fail

Signed-off-by: Christopher Meis <christopher.meis@9elements.com>

### CODE REVIEW

File: cmds/core/cat/cat_test.go [hunk 1]
```diff
@@ -59,3 +68,53 @@ func TestCat(t *testing.T) {
...
+
+	ioCopy = func(out io.Writer, in io.Reader) (int64, error) {
+		return 0, fmt.Errorf("io copy fail")
+	}
+
+	emptyargs := make([]string, 0)
+	if err := cat(&out, emptyargs); !strings.Contains(err.Error(), "io copy fail") {
+		t.Error(err)
+	}
+	resetGlobals()
```
Comment (rminnich) [inline cmds/core/cat/cat_test.go:119]: This is very racy. the resetGlobals of one test might somehow run right after another test has set them. In general, use of this kind of global is a danger sign that things must be done differently. I suggest making the cat function as I described above; and see how some of our other tests run the program and gather the output.

File: cmds/core/cat/cat_test.go [hunk 2]
```diff
@@ -59,3 +68,53 @@ func TestCat(t *testing.T) {
...
+
+func TestFileNotExist(t *testing.T) {
+	args := make([]string, 0)
+	filepath := "testdata/doesnotexist.txt"
+	args = append(args, filepath)
+	var b bytes.Buffer
+	if err := cat(&b, args); !errors.Is(err, os.ErrNotExist) {
+		t.Error(err)
+	}
+	t.Log(b.String())
```
Comment (rminnich) [inline cmds/core/cat/cat_test.go:99]: A bit more informative message here will be useful. Think of the person looking at a log of this in 5 years: what would be most helpful to them? You might even just remove line 99 entirely, since the string appearing in the log might be confusing.

File: cmds/core/cat/cat_test.go [hunk 3]
```diff
@@ -59,3 +68,53 @@ func TestCat(t *testing.T) {
...
+
+	var out bytes.Buffer
+	osStdout = &out
+
+	emptyargs := make([]string, 0)
+	if err := cat(&out, emptyargs); err != nil {
+		t.Error(err)
+	}
+	if out.String() != teststring {
+		t.Errorf("Input string: %s and output: '%s' string are not equal", teststring, out.String())
```
Comment (rminnich) [inline cmds/core/cat/cat_test.go:86]: If you print them with %q it will do the '' for you
If you can follow the standard pattern
t.Errorf("CatPipe: got %q, want %q", teststring, out.String()), would be nice
overall I'm a bit uncomfortable with the globals, it requires the resetGlobal function.

The u-root test framework supports spawning processes and running them. I'd like to hear from Ryan and get his thoughts, but think I'd prefer two things: 
func cat should take an io.Writer and io.Reader... and foreach io.Reader copy it to the io.Writer

For simple tests, the test can call cat function; for full tests, the tests should spawn cat processes and test that way, rather than calling the cat function

File: cmds/core/cat/cat.go [hunk 4]
```diff
@@ -48,12 +61,6 @@ func cat(w io.Writer, files []string) error {
 func main() {
 	flag.Parse()
 
-	if flag.NArg() == 0 {
```
Comment (rminnich) [inline cmds/core/cat/cat.go:51]: I would suggest keeping arg parsing in main and calling cat as a process when you wish to do whole program testing.

File: cmds/core/cat/cat.go [hunk 5]
```diff
@@ -36,8 +37,20 @@ func catFile(w io.Writer, file string) error {
 	return err
 }
 
-func cat(w io.Writer, files []string) error {
-	for _, name := range files {
+var (
+	osStdin  io.ReadWriter = os.Stdin
+	osStdout io.ReadWriter = os.Stdout
```
Comment (rjoleary) [inline cmds/core/cat/cat.go:42]: Instead of global variables, use the pattern like https://github.com/u-root/u-root/blob/master/cmds/contrib/spidev/spidev.go#L113

File: cmds/core/cat/cat_test.go (hunk 2 above)
Comment (ChriMarMe) [reply cmds/core/cat/cat_test.go:99]: Whoops. This really shouldn't be here

**Comment by ChriMarMe** [comment]
Reworked it again. Maybe this is a little better :)

File: cmds/core/cat/cat.go [hunk 6]
```diff
@@ -25,36 +26,51 @@ var (
 	_ = flag.Bool("u", false, "ignored")
 )
 
-func catFile(w io.Writer, file string) error {
-	f, err := os.Open(file)
-	if err != nil {
-		return err
+func catFile(w io.Writer, files []*os.File) error {
```
Comment (rjoleary) [inline cmds/core/cat/cat.go:29]:
```suggestion
func catMultiple(w io.Writer, readers []io.ReadCloser) error {
```

File: cmds/core/cat/cat.go [hunk 7]
```diff
@@ -25,36 +26,51 @@ var (
...
+		writer = os.Stdout
+	}
+	if len(args) == 0 {
+		if err := cat(reader, writer); err != nil {
+			return fmt.Errorf("error concatenating stdin to stdout: %v", err)
+		}
+	}
+	files := make([]*os.File, 0)
+	for _, file := range args {
+		f, err := os.Open(file)
```
Comment (rjoleary) [inline cmds/core/cat/cat.go:59]: This file should be closed somewhere.

File: cmds/core/cat/cat.go [hunk 8]
```diff
@@ -25,36 +26,51 @@ var (
...
+	}
+	if writer == nil {
+		writer = os.Stdout
+	}
+	if len(args) == 0 {
+		if err := cat(reader, writer); err != nil {
+			return fmt.Errorf("error concatenating stdin to stdout: %v", err)
+		}
+	}
+	files := make([]*os.File, 0)
```
[the remark this answers is not in the document]
Comment (rjoleary) [reply cmds/core/cat/cat.go:57]: Also prefer to use the interface `io.ReadCloser`

File: cmds/core/cat/cat.go [hunk 9]
```diff
@@ -25,36 +26,51 @@ var (
...
 )
 
-func catFile(w io.Writer, file string) error {
-	f, err := os.Open(file)
-	if err != nil {
-		return err
+func catFile(w io.Writer, files []*os.File) error {
+	for _, file := range files {
+		if _, err := io.Copy(w, file); err != nil {
+			return fmt.Errorf("error concatenating stdin to stdout: %v", err)
```
Comment (rjoleary) [inline cmds/core/cat/cat.go:32]: This error message is not 100% correct because the reader might not necessarily be stdin.

File: cmds/core/cat/cat.go [hunk 10]
```diff
@@ -25,36 +26,51 @@ var (
 	_ = flag.Bool("u", false, "ignored")
 )
 
-func catFile(w io.Writer, file string) error {
-	f, err := os.Open(file)
-	if err != nil {
-		return err
+func catFile(w io.Writer, files []*os.File) error {
+	for _, file := range files {
+		if _, err := io.Copy(w, file); err != nil {
```
Comment (rjoleary) [inline cmds/core/cat/cat.go:31]: You can re-use the `cat` function here.

**Review by rjoleary — APPROVED** [review]
